File: exercise-5/src/models/Transaction.ts
import Client from "./Client";

class Transaction {
  public income({ client, value }: { client: Client; value: number }): void {
    client.setBalance(client.getBalance() + value);
  }

  public outcome({ client, value }: { client: Client; value: number }): void {
    if (!this.validOutcomeTransaction({ client, value })) {
      console.log("invalid outcome");
      return;
    }

    client.setBalance(client.getBalance() - value);
  }

  public transfer({
    value,
    to,
    from,
  }: {
    value: number;
    to: Client;
    from: Client;
  }): void {
    if (!this.validOutcomeTransaction({ client: from, value })) {
      console.log("invalid outcome");
      return;
    }

    this.outcome({ client: from, value });
    this.income({ client: to, value });
  }

  private validOutcomeTransaction({
    client,
    value,
  }: {
    client: Client;
    value: number;
  }): boolean {
    return client.getBalance() >= value ? true : false;
  }
}

export default Transaction;

File: exercise-5-java/src/com/fapce/plp/models/Transaction.java
package com.fapce.plp.models;

public class Transaction {

    public void income(Client client, Double value) {
        client.setBalance(client.getBalance() + value);
    }

    public void outcome(Client client, Double value) {
        if (validTransaction(client, value)) {
            client.setBalance(client.getBalance() - value);
        } else {
            System.out.println("Invalid outcome.");
        }
    }

    public void transfer(Double value, Client from, Client to) {
        if (validTransaction(from, value)) {
            this.outcome(from, value);
            this.income(to, value);
        } else {
            System.out.println("Invalid transfer.");
        }
    }

    private boolean validTransaction(Client client, Double value) {
        if (client.getBalance() >= value) {
            return true;
        } else {
            return false;
        }
    }

}
